Address-Book.go

```diff
@@ -0,0 +1,348 @@
+		return
+	}
+
+	err = bcrypt.CompareHashAndPassword([]byte(databasePassword), []byte(password))
+	// If wrong password redirect to the login
+	if err != nil {
+		fmt.Println("wrong password")
+		http.Redirect(w, r, "/", http.StatusTemporaryRedirect)
+		return
+	} else {
+		fmt.Println("password match")
+		// If the login succeeded
+		session , _ := store.Get(r,"CurrentSession")
+		session.Values["user"]=username
+		session.Save(r,w)
+		http.Redirect(w, r, "/userpage", 301)
+		return
+	}
+}
+
+
+func Delete(w http.ResponseWriter, r *http.Request){
+	_ ,err := db.Exec("delete from contact where contactID = ?",r.FormValue("id"))
+
+	if err !=nil{
+		fmt.Println("DB error")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	fmt.Println("row Deleted")
+}
+
+
+func DeleteNum(w http.ResponseWriter, r *http.Request){
+	fmt.Println(r.FormValue("id"))
+	_ ,err := db.Exec("delete from phonenumbers where id = ?",r.FormValue("id"))
+
+	if err !=nil{
+		fmt.Println("DB error")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	fmt.Println("number Deleted")
+}
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+func Check(w http.ResponseWriter, r *http.Request) {
+	session , _ := store.Get(r,"CurrentSession")
+	Usr :=session.Values["user"].(string)
+	if Usr !="" {
+		fmt.Println("lesa")
+		http.Redirect(w,r,"/userpage",http.StatusFound)
+		return
+	}else {
+		fmt.Println("logged out")
+		http.Redirect(w,r,"/home",http.StatusFound)
+		return
+		}
+	}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+func HomePage(w http.ResponseWriter, r *http.Request) {
+	fmt.Println("home")
+	templates := template.Must(template.ParseFiles("index.html"))
+	if err := templates.ExecuteTemplate(w, "index.html", nil); err != nil {
+		fmt.Println("error home")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+
+}
+///////////////////////////////////////////////////////////////////////////////////////////////////////
+
+
+func UserPage(w http.ResponseWriter, r *http.Request) {
+	templates := template.Must(template.ParseFiles("userpage.html"))
+	User.Contacts = []Contact{}
+	session , _ := store.Get(r,"CurrentSession")
+	UsernameIn :=session.Values["user"]
+	Username :=UsernameIn.(string)
+	if Username==""{
+		http.Redirect(w,r,"/home",http.StatusFound)
+		return
+	}
+
+	User.UserName =Username
+	row := db.QueryRow("select id from users where username= ?",Username)
+
+	row.Scan(&User.Id)
+	 rows, err := db.Query("select contactID,fname,lname,email from contact where userID= ?",User.Id)
+	if err!=nil{
+		fmt.Println("DB error")
+		http.Error(w,err.Error(),http.StatusInternalServerError)
+
+	}
+
+	for rows.Next() {
+		var c Contact
+		rows.Scan(&c.Id, &c.FirstName, &c.LastName , &c.Email )
+
+		res, err := db.Query("select phonenumber,id from phonenumbers where contact_id= ?",c.Id)
```

Comment:
We should use joins to do these queries. This way instead of calling the database (1 + M + M * N) times, we will call it one time.